lib/pengine/pe_digest.c

```diff
@@ -242,27 +242,28 @@ calculate_secure_digest(op_digest_cache_t *data, pe_resource_t *rsc,
         pcmk__xe_remove_matching_attrs(data->params_secure, attr_in_string,
                                        (void *) secure_list);
     }
-    if (pcmk_is_set(pcmk_get_ra_caps(class),
-                    pcmk_ra_cap_fence_params)) {
+    if (compare_version(op_version, "3.16.0") < 0
```

Comment:
This is an interesting dilemma. op_version will always be the feature set of the controller that recorded the history entry, so that doesn't help us here. We could check the DC version from rsc->cluster->input here instead; that would work ... until the DC is upgraded and we still have older history entries recorded before the upgrade.

What we really want to know is what the DC version was when the history entry was recorded. Interestingly, op_version here is only used to pass to the calculate\_\*\_digest() functions, which only use it to pass to calculate\_operation\_digest(), which doesn't use it at all. I'm wondering if it might be sufficient (and more sensible) for the controller to add the *DC's* version to the history entry instead of its own. Then checking op_version here would be correct. I didn't check whether that's doable.

Reply:
It's kind of confusing since XML_ATTR_CRM_VERSION is added at multiple places. But actually it's the DC version that eventually gets into lrm_rsc_op, which is originally added by the DC when creating a graph action:

https://github.com/ClusterLabs/pacemaker/blob/main/lib/pacemaker/pcmk_graph_producer.c#L332

It is also consistent with the description for cmk__create_history_xml():

https://github.com/ClusterLabs/pacemaker/blob/main/lib/pacemaker/pcmk_sched_actions.c#L891

And by calculating the digests based on the specific DC version indicated in the lrm_rsc_op and comparing against that, it prevents unneeded restart/reload of the resource until it's really necessary.

Reply:
Ah, perfect!